File: 4-Behavioral Patterns/5-Chain of Responsibility Pattern/Usage 2-One or More Receiver Handle Request/Python/atm/atm_machine.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
ATM module.
"""

__author__ = 'Ziang Lu'

from atm.dispatcher import (
    FiveDispatcher, HundredDispatcher, OneDispatcher, TenDispatcher,
    TwentyDispatcher
)


class AtmMachine(object):
    __slots__ = ['_first_dispatcher']

    def __init__(self):
        """
        Default constructor.
        """
        self._first_dispatcher = HundredDispatcher.get_instance(
            TwentyDispatcher.get_instance(
                TenDispatcher.get_instance(
                    FiveDispatcher.get_instance(OneDispatcher.get_instance())
                )
            )
        )

    def withdraw(self, requested_amount: int) -> None:
        """
        Withdraws the given amount of money from this ATM.
        :param requested_amount: int
        :return: None
        """
        # Delegate to the dispatchers to handle this withdraw request
        self._first_dispatcher.dispatch(requested_amount)

File: 4-Behavioral Patterns/5-Chain of Responsibility Pattern/Usage 2-One or More Receiver Handle Request/Java/atm/AtmMachine.java
package atm;

import atm.dispatcher.DollarDispatcher;
import atm.dispatcher.FiveDispatcher;
import atm.dispatcher.HundredDispatcher;
import atm.dispatcher.OneDispatcher;
import atm.dispatcher.TenDispatcher;
import atm.dispatcher.TwentyDispatcher;

/**
 * ATM machine class.
 *
 * @author Ziang Lu
 */
public class AtmMachine {

    /**
     * First dispatcher to handle withdraw requests.
     */
    private final DollarDispatcher firstDispatcher;

    /**
     * Default constructor.
     */
    public AtmMachine() {
        // Chain the dispatchers, which is the chain of responsibility
        firstDispatcher = HundredDispatcher.getInstance(
                TwentyDispatcher.getInstance(
                        TenDispatcher.getInstance(FiveDispatcher.getInstance(OneDispatcher.getInstance()))
                )
        );
    }

    /**
     * Withdraws the given amount of money from this ATM.
     * @param requestedAmount requested amount of money
     */
    public void withdraw(int requestedAmount) {
        // Delegate to the dispatchers to handle this withdraw request
        entryDispatcher.dispatch(requestedAmount);
    }

}
